src/java/org/apache/cassandra/cql3/StatementSource.java

```diff
@@ -0,0 +1,162 @@
+    {
+        if (this == INTERNAL)
+        {
+            return "<<<internal statement>>>";
+        }
+        else
+        {
+            if (bounds != Bounds.INTERNAL && !bounds.isEmpty() && text != null && mayIncludeSourceText)
+                return String.format("at %d:%d - %s", bounds.startLine + 1, bounds.startChar + 1, text);
+            else if (bounds != Bounds.INTERNAL && !bounds.isEmpty())
+                return String.format("at %d:%d", bounds.startLine + 1, bounds.startChar + 1);
+            else if (text != null && !bounds.isEmpty() && mayIncludeSourceText)
+                return String.format("%s", text);
+            else
+                return "";
+        }
+    }
+
+    public static Bounds createBounds(Token startToken, Token endToken)
+    {
+        Objects.requireNonNull(startToken);
+        Objects.requireNonNull(endToken);
+
+        int startLine, startChar, endLine, endChar;
+
+        if (startToken.getType() == Token.EOF)
+        {
+            startLine = Character.MAX_VALUE;
+            startChar = Character.MAX_VALUE;
+        }
+        else
+        {
+            startLine = Math.min(Math.max(startToken.getLine(), 1) - 1, Character.MAX_VALUE);
+            startChar = Math.min(Math.max(startToken.getCharPositionInLine(), 0), Character.MAX_VALUE);
+        }
+
+        if (endToken.getType() == Token.EOF)
+        {
+            endLine = Character.MAX_VALUE;
+            endChar = Character.MAX_VALUE;
+        }
+        else
+        {
+            endLine = Math.min(Math.max(endToken.getLine(), 1) - 1, Character.MAX_VALUE);
+            endChar = Math.min(Math.max(endToken.getCharPositionInLine(), 0), Character.MAX_VALUE);
+        }
+
+        return new Bounds(startLine, startChar, endLine, endChar);
+    }
+
+    public static class Bounds
+    {
+        public static final Bounds INTERNAL = new Bounds(0, 0, Character.MAX_VALUE, Character.MAX_VALUE);
+
+        private static final Pattern LINE_BREAK = Pattern.compile("\\R");
+        private static final String SEPARATOR = " ";
+
+        private final int startLine;
+        private final int startChar;
+        private final int endLine;
+        private final int endChar;
+
+        private Bounds(int startLine, int startChar, int endLine, int endChar)
+        {
+            Preconditions.checkArgument(startLine >= 0 && startLine <= Character.MAX_VALUE);
+            Preconditions.checkArgument(startChar >= 0 && startChar <= Character.MAX_VALUE);
+            Preconditions.checkArgument(endLine >= 0 && endLine <= Character.MAX_VALUE);
+            Preconditions.checkArgument(endChar >= 0 && endChar <= Character.MAX_VALUE);
+
+            this.startLine = startLine;
+            this.startChar = startChar;
+            this.endLine = endLine;
+            this.endChar = endChar;
+        }
+
+        private boolean isEmpty()
+        {
+            return startLine == Character.MAX_VALUE ||
+                   startLine > endLine ||
```

Comment:
what is the case that this happens?  this feels like a bug, should we not fail?

Reply:
In all those pointed out cases I tried to make this class super resistant to wrong input regardless of the circumstances. But it can fail as well.